### PR TITLE
In Summary Tables dialog, prevent error message triggered by invalid R code

### LINKED ISSUE #7446: Error Message when running code for the Summary table
When running the Summary table code, I seem to be getting an error that was previously not there.
@lilyclements I wonder whether this could this be from the changes we have made in the `summary_ table` function?
Am using the survey dataset in this illustration
image

image

image

### PR DESCRIPTION
**Pull request by lilyclements**
Fixes #7446 

@africanmathsinitiative/developers this is ready to review

### CODE REVIEW

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@lilyclements Thank you for this.
- In several places you commented out some unused code. Normally unused code should just be deleted. If there is an exceptional reason to comment it instead, then the comment should include a rationale.
- The PR title will appear in the release notes and is therefore intended for users not developers. Is it possible to change the PR title so that it explains which user visible behaviour has been changed?
- This PR should also be reviewed by someone with R experience. Could you suggest someone?

thanks

**Comment by lilyclements** [comment]
@lloyddewit thanks!

The summary_tables code was changed on two PRs at the same time. Each time a bug was reported, I made the changes on a separate branch with the intention for these changes to be merged into the master branch, and then the PRs to be updated, but for whatever reason this didn’t occur. 
This then caused a conflict and introduced some bugs which spent a bit of time working out.
The unused code was something that I’m not confident we don’t want completely, but I’m not sure when it was relevant. It now introduces a bug. So I commented it out for the time being, with the intention to remove the comment out in the future! (I did of course look for the bug it was related to. It’s very likely that bug was fixed when making changes to the code on the other PR).

I am not very good at writing (or reading) user friendly titles. I think my issue is not knowing the users level of knowledge. Would a user friendly title never include the function name or dlg name? 
Where would I put that information?

**Comment by lilyclements** [comment]
@Ivanluv found the bug so seems suitable to review.
Additionally @HawardKetoyoMsatsi has good experience in R.
So @Ivanluv @HawardKetoyoMsatsi this is ready to review.

**Comment by Ivanluv** [comment]
@lilyclements @lloyddewit  I have been testing the Current updated master and the issue seems to be fixed from my previous PR #7442

**Comment by lloyddewit** [comment]
@lilyclements Thanks for feedback

> I am not very good at writing (or reading) user friendly titles. I think my issue is not knowing the users level of knowledge. Would a user friendly title never include the function name or dlg name?

I think it's good to include the dialog name, possibly the R code function name (if applicable) but not **how** the issue was solved in the code. I made a suggested change to the PR title above.

> Where would I put that information?

All the information that helps developers should be in the PR's opening comment.

**Comment by lilyclements** [comment]
@lloyddewit this branch fixes other bugs, for example, when summarising date variables.

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 1]
```diff
@@ -1360,12 +1360,13 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
     levels(cell_values[[i]]) <- c(levels(cell_values[[i]]), na_level_display)
     cell_values[[i]][is.na(cell_values[[i]])] <- na_level_display
   }
-  grps <- nrow(cell_values)
-  cell_values <- reshape2:::melt.data.frame(cell_values, id.vars = factors, variable.name = "summary-variable", value.name = "value")
+  cell_values <- cell_values %>% dplyr::mutate(dplyr::across(where(is.numeric), round, signif_fig))
+  cell_values <- cell_values %>%
+    tidyr::pivot_longer(cols = !factors, names_to = "summary-variable", values_to = "value", values_transform = list(value = as.character))
+  #cell_values <- reshape2:::melt.data.frame(cell_values, id.vars = factors, variable.name = "summary-variable", value.name = "value")
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/Backend_Components/summary_functions.R:1366]:
```suggestion
```

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 2]
```diff
@@ -1458,22 +1459,38 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
             # TODO: if the colname is the same as a factor, then do nothing
             colnames(summary_margins)[col] <- sub("_value", "_all", colnames(summary_margins)[col])
           }
+          summary_margins <- summary_margins %>% dplyr::mutate(dplyr::across(where(is.numeric), round, signif_fig))
           summary_margins <- summary_margins %>%
-          tidyr::pivot_longer(cols = !factors, names_to = "summary-variable", values_to = "value")
+            tidyr::pivot_longer(cols = !factors, names_to = "summary-variable", values_to = "value", values_transform = list(value = as.character))
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/Backend_Components/summary_functions.R:1464]:
```suggestion
          tidyr::pivot_longer(cols = !factors, names_to = "summary-variable", values_to = "value", values_transform = list(value = as.character))
```

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 3]
```diff
@@ -1458,22 +1459,38 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
...
+                             i, ~ forcats::fct_relevel(., shaped_cell_values_levels))
+        }     
+      }
       shaped_cell_values <- dplyr::bind_rows(shaped_cell_values, margin_tables_all) %>%
-        dplyr::mutate_at(vars(-value), ~ replace(., is.na(.), margin_name)) %>%
-        dplyr::mutate_at(vars(-value), ~ forcats::as_factor(forcats::fct_relevel(., margin_name, after = Inf)))
+        dplyr::mutate_at(vars(-c(value)), tidyr::replace_na, margin_name) %>%
+        dplyr::mutate_at(vars(-c(value)), ~forcats::as_factor(forcats::fct_relevel(.x, margin_name, after = Inf)))
+      
+
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/Backend_Components/summary_functions.R:1493]:
```suggestion
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]

**Review by lloyddewit — DISMISSED** [review]

**Comment by lloyddewit** [comment]
@lilyclements I made some small cosmetic changes, please check that you are OK with them

@rdstern If you can test/approve, then we can merge, thanks

**Comment by lilyclements** [comment]
@lloyddewit thanks for making these changes - look good!

**Comment by rdstern** [comment]
I am not sure what's happening here.  I get an error of the same sort that @Ivanluv indicates in the issue.  But thankfully Version 0.7.5 seems to be working ok.  Here is an example to compare the code:  First from 0.7.5
# Code generated by the dialog, Summary\Frequency Tables
summary_table <- data_book$summary_table(data_name="survey", columns_to_summarise="yield", factors=c("village","variety"), treat_columns_as_factor=FALSE, summaries=c("summary_mean"))


last_table <- (mmtable2::mmtable(data=summary_table, cells=value) + mmtable2::header_top_left(variable='summary-variable') + mmtable2::header_top_left(variable=village) + mmtable2::header_left_top(variable=variety))
data_book$add_table(table_name="last_table", table=last_table, data_name="survey")
data_book$get_tables(data_name="survey", table_name="last_table")

And from the latest pull request I get:
image
From 
image

with code:
# Code generated by the dialog, Summary Tables
summary_table <- data_book$summary_table(data_name="Sheet1", columns_to_summarise="yield", factors=c("village","variety"), treat_columns_as_factor=FALSE, summaries=c("summary_mean"))


last_table <- (mmtable2::mmtable(data=summary_table, cells=value) + mmtable2::header_top_left(variable='summary-variable') + mmtable2::header_top_left(variable=village) + mmtable2::header_left_top(variable=variety))
data_book$add_table(table_name="last_table", table=last_table, data_name="Sheet1")
data_book$get_tables(data_name="Sheet1", table_name="last_table")


And I can't see the difference?  Confusing - at least to me!

**Comment by lilyclements** [comment]
@rdstern the only difference I can see is that the former uses `data_name = "survey"`, and the later `data_name = "sheet1"`. Is the data called survey in both runs?

**Comment by rdstern** [comment]
@lilyclements it was called survey and came from the library in 0.7.5 and sheet1 and came from my own copy for the new version.  I don't see why that's caused a problem.  perhaps we look together tomorrow?  We may have quite a full day!

**Comment by lilyclements** [comment]
@rdstern I looked into this a bit more, and saw that the output is all `NA` from the summary statistics:

```
summary_table <- data_book$summary_table(data_name="survey", columns_to_summarise="yield", factors=c("village","variety"), treat_columns_as_factor=FALSE, summaries=c("summary_mean"))

head(summary_table)
# A tibble: 6 x 4
# Groups:   village, variety [6]
  village variety `summary-variable` value
  <fct>   <fct>   <chr>              <dbl>
1 KESEN   OLD     mean__yield           NA
2 KESEN   TRAD    mean__yield           NA
3 NANDA   NEW     mean__yield           NA
4 NANDA   OLD     mean__yield           NA
5 NANDA   TRAD    mean__yield           NA
6 NIKO    OLD     mean__yield           NA
```

I believe the error here is occurring due to all the `NA` values. I guess we would expect essentially an empty table from `mmtable2`. What should we do about this? Is it sensible to report it to the package author?

**Comment by lilyclements** [comment]
@rdstern apologies - `yield` should not be giving `NA`'s in the output! I'll look into that now

**Comment by lilyclements** [comment]
@rdstern I have got to the bottom of the problem! If we make the `summary_table` object a tibble or data frame then it works fine.  I have now fixed this in the R-code. Can you recheck and let me know if it is still an issue for you? It is now fixed my end.

**Comment by rdstern** [comment]
@lilyclements it now seems to work fine for frequency tables and for summary tables with one factor.  In a summary table with both village and variety it still gives an error.

image

I wondered if it was the gaps in the results, but it  still gives an error with just the 2 villages, where all 3 varieties are used.

**Comment by lilyclements** [comment]
@rdstern can you send a copy of the script? My script is the following and I am not able to replicate this error. I'll get to the bottom of it soon I'm sure,

```
# Setting working directory, sourcing R code and loading R packages
setwd(dir="C:/Users/lclem/source/repos/R-Instat/instat/bin/Debug/static/InstatObject/R")

source(file="Rsetup.R")

data_book <- DataBook$new()

options(dplyr.summarise.inform=FALSE)

# Option: Number of digits to display
options(digits=4)

# Option: Show stars on summary tables of coefficients
options(show.signif.stars=FALSE)

# Code generated by the dialog, Import Dataset
new_RDS <- readRDS(file="C:/Users/lclem/OneDrive - University of Southampton/PhD/RInstat/Data/experimental_survey.RDS")
data_book$import_RDS(data_RDS=new_RDS)

# Code generated by the dialog, Summary\Frequency Tables
summary_table <- data_book$summary_table(data_name="survey", columns_to_summarise="yield", factors=c("village","variety"), treat_columns_as_factor=FALSE, summaries=c("summary_mean"))

last_table <- (mmtable2::mmtable(data=summary_table, cells=value) + mmtable2::header_top_left(variable='summary-variable') + mmtable2::header_top_left(variable=village) + mmtable2::header_left_top(variable=variety))
data_book$add_table(table_name="last_table", table=last_table, data_name="survey")
data_book$get_tables(data_name="survey", table_name="last_table")
```

**Comment by rdstern** [comment]
Here is the error:
image

```
setwd(dir="C:/Users/RogerStern/source/repos/africanmathsinitiative/R-Instat/instat/bin/x64/Debug/static/InstatObject/R")

source(file="Rsetup.R")

data_book <- DataBook$new()

options(dplyr.summarise.inform=FALSE)

# Option: Number of digits to display
options(digits=4)

# Option: Show stars on summary tables of coefficients
options(show.signif.stars=FALSE)

# Code generated by the dialog, Import Dataset
Sheet1 <- rio::import(file="C:/Users/RogerStern/SSD Dropbox/Roger Stern/R-Instat Documentation/Data/rice_survey.xlsx", guess_max=Inf, which=1)
data_book$import_data(data_tables=list(Sheet1=Sheet1))


rm(Sheet1)

# Right click menu: Convert Column(s) To Factor
data_book$convert_column_to_type(data_name="Sheet1", col_names="village", to_type="factor")

# Right click menu: Convert Column(s) To Factor
data_book$convert_column_to_type(data_name="Sheet1", col_names="variety", to_type="factor")

# Right click menu: Convert Column(s) To Factor
data_book$convert_column_to_type(data_name="Sheet1", col_names="fertgrp", to_type="factor")

# Code generated by the dialog, Summary Tables
summary_table <- data_book$summary_table(data_name="Sheet1", columns_to_summarise="yield", factors=c("village","variety"), treat_columns_as_factor=FALSE, summaries=c("summary_mean"))


last_table <- (mmtable2::mmtable(data=summary_table, cells=value) + mmtable2::header_top_left(variable='summary-variable'))
data_book$add_table(table_name="last_table", table=last_table, data_name="Sheet1")
data_book$get_tables(data_name="Sheet1", table_name="last_table")

rm(list=c("last_table", "summary_table"))

```

**Comment by lilyclements** [comment]
Thanks @rdstern - I am still not getting the error. Can you send me a copy of your data? Perhaps a variable class is different etc!

**Comment by lilyclements** [comment]
Good to know. I am still not getting an error. Final one - can you try updating this branch? I've now merged the master into it. If it's not that, then we can look at it tomorrow.

**Comment by lilyclements** [comment]
Great! Glad we got there.

**Comment by lilyclements** [comment]
@rdstern are you happy to approve these changes?

**Review by rdstern — APPROVED** [review]
Great

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 4]
```diff
@@ -1385,20 +1385,25 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
     for (facts in power_sets_outer) {
       if (length(facts) == 0) facts <- c()
       margin_tables[[length(margin_tables) + 1]] <- self$calculate_summary(data_name = data_name, columns_to_summarise = columns_to_summarise, summaries = summaries, factors = facts, store_results = FALSE, drop = drop, na.rm = na.rm, return_output = TRUE, weights = weights, result_names = result_names, percentage_type = percentage_type, perc_total_columns = perc_total_columns, perc_total_factors = perc_total_factors, perc_total_filter = perc_total_filter, perc_decimal = perc_decimal, margin_name = margin_name, additional_filter = additional_filter, perc_return_all = FALSE, sep = "__", ...)
-      margin_tables[[length(margin_tables)]] <- margin_tables[[length(margin_tables)]] %>% dplyr::select(c(all_of(facts), order_names))
+      margin_tables[[length(margin_tables)]] <- margin_tables[[length(margin_tables)]] #%>% dplyr::select(c(tidyselect::all_of(facts), tidyselect::all_of(order_names)))
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/Backend_Components/summary_functions.R:1388]: what is the purpose of this line?

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@lilyclements Thanks, just one question

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R (hunk 4 above)
Comment (lilyclements) [reply instat/static/InstatObject/R/Backend_Components/summary_functions.R:1388]: @lloyddewit good point - I suppose this can be removed

**Review by lloyddewit — APPROVED** [review]

**Pull request merged**